Chatbot checkboxes disabled on submit

katieho-va opened the pull request:
## Description
- Refactors event listener logic around disabling buttons, disabling checkboxes (new!), and scrolling to the newest user message
- Moved GA events around, refactors order @batemapf 
- Fixes this bug for chatbot checkboxes [department-of-veterans-affairs/covid19-chatbot#138](https://github.com/department-of-veterans-affairs/covid19-chatbot/issues/138) @pjhill 
- Extracts `window.WebChat.renderWebChat(webchatOptions, root)` call from within `index.js/startChat` method

## Testing done
Just manual testing for now. Works in Chrome, Safari, Firefox, IE11.

**Next steps: an automated UI test to validate this functionality.**

## Screenshots
<img width="550" alt="Screen Shot 2020-04-23 at 4 59 22 PM" src="https://user-images.githubusercontent.com/60353062/80149412-9cbc8800-8584-11ea-9e59-f982f4c514a5.png">

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

katieho-tw committed:
[department-of-veterans-affairs/covid19-chatbot#138] Disables checkboxes when submit button is clicked

katieho-tw committed:
[department-of-veterans-affairs/covid19-chatbot#138] Extracts renderWebChat call to index.js

Moves recordEvents on connection success/failure to index.js

Co-authored-by: Shefali Nayak <shefali.nayak@thoughtworks.com>

katieho-tw committed:
[department-of-veterans-affairs/covid19-chatbot#138] Refactors event listener logic

Co-authored-by: Shefali Nayak <shefali.nayak@thoughtworks.com>

katieho-tw committed:
[department-of-veterans-affairs/covid19-chatbot#138] Removes thrown error and eslint-disable

katieho-tw committed:
[department-of-veterans-affairs/covid19-chatbot#170] Refactors order of promises and combines calls to recordEvent

Co-authored-by: Shaun Hubbard <shaun.hubbard@thoughtworks.com>

katieho-va commented:
The [jenkins failure](http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-website/detail/checkboxes-disabled-on-submit/1/pipeline) above points out some AXE accessibility failures, but since we are using this botframework for the chatbot, we don't have access to that html. @1Copenut FYI

ncksllvn commented on src/applications/coronavirus-chatbot/createCoronavirusChatbot.js:
```diff
@@ -10,23 +10,30 @@ export default (_store, widgetType) => {
-      recordEvent({
-        event: `${GA_PREFIX}-load-failure`,
-        'error-key': undefined,
+  import(/* webpackChunkName: "chatbot" */ './index').then(module => {
+    const initializeChatbot = module.default;
+    initializeChatbot()
+      .then(webchatOptions => {
+        recordEvent({
+          event: `${GA_PREFIX}-connection-successful`,
+          'error-key': undefined,
+        });
+        recordEvent({
+          event: `${GA_PREFIX}-load-successful`,
+          'error-key': undefined,
+        });
+
+        window.WebChat.renderWebChat(webchatOptions, root);
+      })
+      .catch(() => {
+        recordEvent({
+          event: `${GA_PREFIX}-connection-failure`,
+          'error-key': 'XX_failed_to_start_chat',
+        });
+        recordEvent({
+          event: `${GA_PREFIX}-load-failure`,
+          'error-key': undefined,
+        });
```
Optionally, you could switch to async/await
```suggestion
  import(/* webpackChunkName: "chatbot" */ './index').then(async module => {
    const initializeChatbot = module.default;
    try {
        const webchatOptions = await initializeChatbot();
        recordEvent({
          event: `${GA_PREFIX}-connection-successful`,
          'error-key': undefined,
        });
        recordEvent({
          event: `${GA_PREFIX}-load-successful`,
          'error-key': undefined,
        });
        window.WebChat.renderWebChat(webchatOptions, root);
    } catch(err) {
      recordEvent({
          event: `${GA_PREFIX}-connection-failure`,
          'error-key': 'XX_failed_to_start_chat',
        });
        recordEvent({
          event: `${GA_PREFIX}-load-failure`,
          'error-key': undefined,
        });
    }
```

ncksllvn commented on src/applications/coronavirus-chatbot/utils.js:
```diff
@@ -17,12 +9,37 @@ const disableButtons = event => {
 };
 
-export const watchForButtonClicks = () => {
+const disableCheckboxes = () => {
+  const inputs = document
+    .getElementById('webchat')
+    .getElementsByTagName('input');
+
+  for (let i = 0; i < inputs.length; i++) {
+    if (inputs[i].type === 'checkbox') {
+      inputs[i].disabled = true;
+    }
+  }
+};
```
You could dodge the for-loop like - 


```suggestion
const disableCheckboxes = () => {
[...document.querySelectorAll('#webchat input[type="checkbox"]')]
    .forEach(input => input.disabled = true);
};
```

katieho-va committed:
Update src/applications/coronavirus-chatbot/createCoronavirusChatbot.js

Co-Authored-By: Nick Sullivan <nick.sullivan@adhocteam.us>

katieho-va committed:
Update src/applications/coronavirus-chatbot/utils.js

Co-Authored-By: Nick Sullivan <nick.sullivan@adhocteam.us>

katieho-tw committed:
[department-of-veterans-affairs/covid19-chatbot#138] Refactors suggestions to use async/await and avoid for-loop for inputs

Mottie commented on src/applications/coronavirus-chatbot/utils.js:
```diff
@@ -17,12 +9,35 @@ const disableButtons = event => {
     for (let i = 0; i < buttons.length; i++) {
-      buttons[i].addEventListener('click', disableButtons);
-      buttons[i].addEventListener('click', scrollToNewMessage);
+      buttons[i].addEventListener('click', handleDisableAndScroll);
```
It might be better to use delegated event binding here, i.e. bind 'click' to a wrapping element, then check to see if the correct button was clicked when the event bubbles up.

katieho-va replied:
Yes, we had a similar suggestion on a previous [PR](https://github.com/department-of-veterans-affairs/vets-website/pull/12298/files) but found it was messier to do in the `botframework` code.

> The event listener on each button (added by botframework-webchat) doesn't bubble the event up so our event listener on root was never getting executed. It seems messier to change the webchat code, so we stuck with applying our event listener dynamically.

1Copenut commented:
@katieho-va Thank you for the heads up!

> The [jenkins failure](http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-website/detail/checkboxes-disabled-on-submit/1/pipeline) points out some AXE accessibility failures, but since we are using this botframework for the chatbot, we don't have access to edit that html. @1Copenut FYI

I took a look at the Jenkins log this morning, and can see the two issues:
1. The `ul[role="list"]` must have child `li[role="listitem"]` elements to pass validation
2. The `div[aria-labelledby="webchat__toaster__header__jcaff"]` is looking for an ID that is dynamically generated, so we can't hard-code that ID into our introductory paragraph as we originally hoped.

@ncksllvn Any idea how we get around these issues? As long as the chatbot is on one page, we may have to exclude that page from axe checks, or at least exclude those two rules for that page.

katieho-va commented:
> @katieho-va Thank you for the heads up!
> 
> > The [jenkins failure](http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-website/detail/checkboxes-disabled-on-submit/1/pipeline) points out some AXE accessibility failures, but since we are using this botframework for the chatbot, we don't have access to edit that html. @1Copenut FYI
> 
> I took a look at the Jenkins log this morning, and can see the two issues:
> 
> 1. The `ul[role="list"]` must have child `li[role="listitem"]` elements to pass validation
> 2. The `div[aria-labelledby="webchat__toaster__header__jcaff"]` is looking for an ID that is dynamically generated, so we can't hard-code that ID into our introductory paragraph as we originally hoped.
> 

@1Copenut For #2, are you referring to this [issue fix](https://github.com/department-of-veterans-affairs/covid19-chatbot/issues/100)?

1Copenut commented:
> @1Copenut For #2, are you referring to this [issue fix](https://github.com/department-of-veterans-affairs/covid19-chatbot/issues/100)?

Yes, that's the one. I hoped the ID was static, but it does appear to be generated by chatbot code dynamically. The ID changed from the issue ticket to the Jenkins logs, so we have to consider that a chatbot issue unless there's a custom prop we can pass in to override the dynamic nature.

katieho-va commented:
> Yes, that's the one. I hoped the ID was static, but it does appear to be generated by chatbot code dynamically. The ID changed from the issue ticket to the Jenkins logs, so we have to consider that a chatbot issue unless there's a custom prop we can pass in to override the dynamic nature.

Ah I see, ok. I'll link this conversation to that issue so we know.

cvalarida reviewed (approved):
I didn't look too closely at the changes, but nothing stood out that would block this if you've tested that it works. :shipit:

ncksllvn commented on src/applications/coronavirus-chatbot/utils.js:
```diff
@@ -17,12 +9,35 @@ const disableButtons = event => {
+const disableCheckboxes = () => {
+  [...document.querySelectorAll('#webchat input[type="checkbox"]')].forEach(
+    input => {
+      const currentInput = input;
```
Lol at this linter-bypass!

katieho-tw committed:
[department-of-veterans-affairs/covid19-chatbot#138] Adds chatbot page to exempted pages for accessibility fixes for now

Most of the accessibility violations flagged by Axe would require edits to the bot-framework code
